blog/tech-talk-window-resize-behavior.md

```diff
@@ -0,0 +1,290 @@
+We still show the old frame in the green viewport. But the clip rect has become larger and we show areas of the surface that we haven't repainted yet.
+
+On the first resize of a window, these areas would be black. On the second resize, those areas would be filled with old pixel values. They would show whatever we had previously painted in those areas.
+
+Similarly, in a certain edge case while making the window smaller, we would sometimes repaint the viewport before we would update the clip rect.
+
+![Visualization where the viewport was repainted before the clip rect was updated](/assets/img/blog/tech-talk-window-resize-behavior/visualization-4.png)
+
+Then parts of the clip rect would still show the previous frame because the new frame was smaller and we did not repaint any areas beyond the new viewport.
+
+Now why do these operations not happen in sync?
+
+We use two different Windows APIs here:
+
+- [`IDXGISwapChain1::Present1`](https://source.chromium.org/chromium/chromium/src/+/main:gpu/command_buffer/service/shared_image/dxgi_swap_chain_image_backing.cc;l=283;drc=92e802d8fa66bafb0cca9ed32143d6148d8e0411) — This shows the new pixels on screen / updates the new viewport.
+- [`IDCompositionDevice::Commit`](https://source.chromium.org/chromium/chromium/src/+/main:ui/gl/dc_layer_tree.cc;l=1034;drc=35d26d364efb57d0386b98312ba739f7f65ae97e) — This updates the clip rect.
+
+Here's what's important to understand: both functions return synchronously on the CPU. However, they schedule tasks that run asynchronously on the GPU at a later time. Windows and its services (such as [DWM](https://en.wikipedia.org/wiki/Desktop_Window_Manager)) decide when they will run and in which order. So they take effect asynchronously, and not always within the same frame.
+
+Unfortunately, Windows provides no way for us to synchronize those operations. So I had to find other approaches to fix this.
+
+There were two options that I evaluated with the Chromium maintainers:
+
+1. While resizing, paint all previously drawn areas outside of the viewport transparent. This makes those areas invisible. It fixes the artifacts.
+2. While resizing, switch from an `IDXGISwapChain1` to a DirectComposition surface which synchronizes updates with `IDCompositionDevice::Commit`. This also fixes the artifacts.
+
+We went with the first option because it leads to faster resizes than the second option.
+
+I landed a [patch](https://crrev.com/c/7115438) in Chromium that implemented that first solution.
+
+I also submitted two other patches in preparation for the main patch:
+
+1. The [first one](https://crrev.com/c/7129658) fixed a bug in existing code that would make CI fail in combination with the main patch. It also made launching Electron apps and Chrome a tiny bit faster.
+2. The [second one](https://crrev.com/c/7210913) was split off to make code review easier on the main patch.
+
+## Second bug
+
+In addition to this first bug, there was a second bug which also led to stale pixels.
```

Comment:
```suggestion
In addition to the `DirectComposition` issue, I discovered a second bug that also led to stale pixels.
```

Active voice!

Reply:
Thanks for the suggestion! To be honest, I'd prefer to keep it as is. In general, I agree that active voice is better. Here I prefer the more neutral phrasing instead of highlighting my achievements (as in "I did this").